File: elixir/dnd-character/lib/dnd_character.ex
defmodule DndCharacter do
  @type t :: %__MODULE__{
          strength: pos_integer(),
          dexterity: pos_integer(),
          constitution: pos_integer(),
          intelligence: pos_integer(),
          wisdom: pos_integer(),
          charisma: pos_integer(),
          hitpoints: pos_integer()
        }

  defstruct ~w[strength dexterity constitution intelligence wisdom charisma hitpoints]a

  @spec modifier(pos_integer()) :: integer()
  def modifier(score) do
    Integer.floor_div(score - 10, 2)
  end

  @spec ability :: pos_integer()
  def ability do
    rolls =
      for _ <- 1..4 do
        Enum.random(1..6)
      end

    Enum.sum(rolls) - Enum.min(rolls)
  end

  @spec character :: t()
  def character do
    character =
      for characteristic <- Map.keys(%__MODULE__{}),
          characteristic != :__struct__,
          reduce: %__MODULE__{}
      do
        char -> %{char | characteristic => ability()}
      end

    %{character | :hitpoints => 10 + modifier(character.constitution)}
  end
end

File: nim/dnd-character/dnd_character.nim
import std/math
import std/random
import sugar

type
  Character* = object
    strength*: int
    dexterity*: int
    constitution*: int
    intelligence*: int
    wisdom*: int
    charisma*: int
    hitpoints*: int

proc ability*: int =
  var d6 = () => rand(1..6)
  var roll = [d6(), d6(), d6(), d6()]
  sum(roll) - min(roll)

proc modifier*(n: int): int =
  floorDiv(n - 10, 2)

proc initCharacter*: Character =
  var constitution = ability()
  Character(
    strength:     ability(),
    dexterity:    ability(),
    intelligence: ability(),
    wisdom:       ability(),
    charisma:     ability(),
    constitution: constitution,
    hitpoints:    10 + modifier(constitution)
  )
